src/main/java/fi/dy/masa/malilib/network/payload/PayloadManager.java

```diff
@@ -0,0 +1,227 @@
+ * Payload Record file that is created.
+ */
+public class PayloadManager
+{
+    private static final PayloadManager INSTANCE = new PayloadManager();
+    public static PayloadManager getInstance() { return INSTANCE; }
+    private final Map<PayloadType, PayloadCodec> TYPES = new HashMap<>();
+
+    public PayloadManager() {}
+
+    /**
+     * Registers a Payload Type with Payload Manager
+     *
+     * @param type (PayloadType ENUM)
+     * @param id (Identifier Namespace:Path, ie, servux:structures)
+     */
+    public void register(PayloadType type, Identifier id)
+    {
+        if (TYPES.containsKey(type) == false && type.exists(type))
+        {
+            PayloadCodec codec = new PayloadCodec(type, id);
+            TYPES.put(type, codec);
+        }
+    }
+
+    /**
+     * Abstract method for CustomPayload's to define their PACKET_CODEC value.
+     */
+    @Nullable
+    public PayloadCodec getPayloadCodec(PayloadType type)
+    {
+        return TYPES.getOrDefault(type, null);
+    }
+
+    /**
+     * Abstract method for CustomPayload's to define their PACKET_TYPE value, derived from the channel Identifier
+     */
+    @Nullable
+    public Identifier getIdentifier(PayloadType type)
+    {
+        return TYPES.getOrDefault(type, null).getId();
+    }
+
+    /**
+     * Search for a registered Payload type by Identifier
+     */
+    @Nullable
+    public PayloadType getPayloadType(Identifier id)
+    {
+        for (PayloadType type : TYPES.keySet())
+        {
+            PayloadCodec codec = TYPES.get(type);
+
+            if (codec != null && codec.getId().equals(id))
+            {
+                return type;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Fabric API Play Channel registration.
+     */
+    public <T extends CustomPayload> void registerPlayChannel(PayloadType type, CustomPayload.Id<T> id, PacketCodec<PacketByteBuf, T> packetCodec)
+    {
+        PayloadCodec codec = this.getPayloadCodec(type);
+
+        // Never Attempt to "re-register" a channel or bad things will happen.  Kittens harmed, etc.
+        if (codec == null || codec.isPlayRegistered())
+        {
+            return;
+        }
+
+        codec.registerPlayCodec();
+
+        // Checks with Fabric APIs IMPL layer (I don't think they are confident with their code yet)
+        if (PayloadTypeRegistryImpl.PLAY_S2C.get(id) != null || PayloadTypeRegistryImpl.PLAY_C2S.get(id) != null)
+        {
+            // This just saved Minecraft from crashing, your welcome.
+            MaLiLib.logger.error("registerPlayChannel: blocked duplicate Play Channel registration attempt for: {}.", id.id().toString());
+        }
+        else
+        {
+            //MaLiLib.logger.info("registerPlayChannel: [Fabric-API] registering Play Channel: {}", id.id().toString());
+            PayloadTypeRegistry.playC2S().register(id, packetCodec);
```

Comment:
check direction to which side it needs to be sent maybe?

Reply:
You need both.

Reply:
And Correction, the register.play C2S/S2C part has to do with inbound data versus outbound for the environment your working in.  You *could* only register a single direction, but then your mod will be limited to "only receiving" or "only sending" data.

Reply:
For Mods like Servux <-> MiniHUD you want it to be bi-directional so that they can handshake data.